contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+
+## Motivation
+From user_namespaces(7):
+> User namespaces isolate security-related identifiers and attributes, in particular, user IDs and group IDs, the root directory, keys, and capabilities. A process's user and group IDs can be different inside and outside a user namespace. In particular, a process can have a normal unprivileged user ID outside a user namespace while at the same time having a user ID of 0 inside the namespace; in other words, the process has full privileges for operations inside the user namespace, but is unprivileged for operations outside the namespace.
+
+In order to run Pods with software which expects to run as root or with elevated privileges while still containing the processes and protecting both the Nodes and other Pods, Linux kernel mechanism of user namespaces can be used make the processes in the Pods view their environment as having the privileges, while on the host (Node) level these processes appear as without privileges or with privileges only affecting processes in the same Pods
+
+The purpose of using user namespaces in Kubernetes is to let the processes in Pods think they run as one uid set when in fact they run as different “real” uids on the Nodes.
+
+In this text, most everything said about uids can also be applied to gids.
+
+## Goals
+- Enable user namespace support in a kubernetes cluster so that workloads that work today also work with user namespaces enabled. Furthermore, make workloads that require root inside the container safer using the additional security of user namespaces. 
+- The goal is to be able to take container images that would all want to run as the same uid (0 for root, 48 for apache, some 997 for nginx) or set of uids (systemd-based multi-daemon container which starts as uid 0 and forks a number of service running as different uids) and run them without clashing with the same uids on the host.
+- The effort will be successful when it will be possible to use any container image, built using any set of uids and run it in Kubernetes cluster safely with host/node  uids not matching those in the image. Specifically, Pods with containers that need to be run as root (uid 0) should be runnable with non-root uids.
+- To allow images using privileged users securely where such containers will run in a user namespace different from user-namespace of the underlying host.  
+
+## Non-Goals
+- Non-goal is to support use cases when the uids for processes get allocated in runtime and externally (IPA, Active Directory) for generic users, without specifying (hopefully narrow) range of uids to be mapped upfront, before the container gets started.
+- Non-goal is to support pod/container level user namespace isolation. There can be images using different users but on the node, pods/containers running with these images will share common user namespace remapping configuration. In other words, all containers on a node share a common user-namespace range.
```

Comment:
We are trying to use user-namespace feature to the level it is being supported in more than one runtimes and not just docker.  No runtime today supports passing a per-container UID range. Even `rkt` chooses a random UID range from host and provides no guarantee that those ranges will not collide. Therefore, we thought of going incrementally in adding layers of security that user-namespaces can provide.
 
With what we have today in runtimes in terms of user-namespace support, through this proposal we are trying to secure nodes from rough pod containers. Once runtimes have evolved enough to be able to allow per-container UID range, we will then design to add another layer of security to protect pods of one namespace from the pods of another namespace.

> the only real change is that root in containers is no longer "real root" but we already have PSPs that can be used to restrict running as root

Thats the point. We dont want to limit and restrict those images which are supposed to run as root user.

Reply:
@cyphar, we see the security benefits in being able to isolate the Node from the Pods in most cases, while being able to for example run systemd-based and multi-daemon containers without allowing root on the host.

While going all-in with per-Pod user namespaces would be good and it was the goal of the previous PR https://github.com/kubernetes/community/pull/2042, we've been advised that the proposal should focus on something that Container Runtimes can actually currently provide. There is an ongoing effort brewing in https://github.com/kubernetes-incubator/cri-o/pull/1519 and I've been following that and was able to specify uid/gid mapping per-Pod when using that branch of CRI-O.

Do you feel that the current proposal hinders future adoption of per-Pod user namespace support? Our hope was that it will actually validate the user namespace functionality and prepare way for future more flexible setups.

Reply:
> No runtime today supports passing a per-container UID range. Even rkt chooses a random UID range from host and provides no guarantee that those ranges will not collide.

LXC **definitely** does support this, and `rkt` might not provide such a guarantee but it doesn't change that they do support it too (and in principle could implement such a guarantee if they wanted to). Not to mention that `runc` and the OCI spec explicitly support this. So I disagree with this conclusion.

To be fair, `cri-o` doesn't support this but from a high-level perspective the implementation would not be *significantly* harder than the current plan of implementing user namespaces (Docker is more difficult for some more nuanced reasons -- not to mention that it is already implemented and changing it after-the-fact is much harder). It however would be significantly harder to retro-fit a pod-specific UID mapping to implementations like `cri-o` (as is evident from how difficult it would be to add that design to Docker).

> Therefore, we thought of going incrementally in adding layers of security that user-namespaces can provide.

I don't have a problem with doing this incrementally, I'm just worried that you cannot incrementally add the design I described above with this design -- you would need to create a separate design to this (and then you have two different ways of setting up user namespaces).

> Thats the point. We dont want to limit and restrict those images which are supposed to run as root user.

But there are many benefits of user namespaces other than just this one feature -- which is the point I'm trying to make. A large portion of container images don't *need* root privileges to run (and user namespaces still are very useful for images which don't run as root -- because it protects against cross-container attacks).

Reply:
@adelton 

> we see the security benefits in being able to isolate the Node from the Pods in most cases, while being able to for example run systemd-based and multi-daemon containers without allowing root on the host.

I agree that this is *a* benefit of user namespaces, but you can (in principle) get the same benefit by not running as root inside a non-user namespaced container. To be fair, there are some additional protections you get from running containers in a user namespace versus just running as an unprivileged user in the host user namespace, but they are *very* specific and nuanced (for instance the `ptrace` permission checks are more rigid and protect against some join-race attacks). I get that the main benefit of this is that it allows people who currently run containers as root to get more security, but I feel that making that *one of the few benefits* of using user namespaces in Kubernetes is restricting ourselves needlessly.

> Do you feel that the current proposal hinders future adoption of per-Pod user namespace support? Our hope was that it will actually validate the user namespace functionality and prepare way for future more flexible setups.

Yes, I believe it might (but I might be mistaken -- I've only skimmed the full design doc, I will re-read it tomorrow).

The design I outlined above -- where you by default have an unspecified user namespace mapping (so it's unique for each container with no overlaps), and for special cases like `hostPath` you specify what UIDs have to be the same in all containers that access that `hostPath` (which allows you to then have only overlaps for those UIDs and only for containers that require such access). Maybe this links into the "named user namespaces" thing that gets mentioned later (though I don't know what the term means), but this design would permit **far more** out-of-the-box security and also provides far more semantic control to users to specify what UIDs they actually want to share.

But the design I outlined requires more changes to the Kubernetes API, so it's *possible* that iteration on this design could be done to implement the design that I outlined (my main concern is whether or not the mappings are entirely controlled by Kubernetes -- because there are a lot of things that are runtime-specific that Kubernetes doesn't really know about and shouldn't care about when it comes to user namespaces).

Reply:
@cyphar The end goal is definitely going towards per pod (or per namespace) user namespaces. However, we feel that we should go there step by step. Even getting this first step working seamlessly in kubernetes would take quite some effort. I think one of the principles we can follow is minimize user burden while enabling user namespaces. They simply shouldn't know or care or should have to modify their workloads or pod/resource settings. They should benefit from the additional security of user namespaces when they are enabled in the cluster. 

For the use case of per-pod user namespace, I think it would make sense to allocate a user namespace to a k8s namespace and then have all pods within that k8s namespaces share the same mappings. I can't think of anything that is being proposed as a first step here that would prevent us to move towards that in the future. But again, there is still a lot of work to do to even get this first step working nicely :)

Reply:
@cyphar thanks a lot and i really appreciate your inputs. Just want to clarify on this:

> . But I'm not entirely sure whether we would end up with two different ways of specifying user namespaces

In our understanding this is not going to be case and we will not end-up specifying user namespaces in two different ways. Today we are reading a single default mapping from runtime and using it for all the pods on a node. In future also, to enable pod (k8s namespace level) user-namespaces, this mapping will be read from runtime and then a sub-part will be used for each pod (or k8s namespace). Externally, there will be no change in the way to enable user-namespaces or configuring user-namespaces. Ofcourse, this future approach still need more rigorous thinking to become a complete approach, but does seem to us that we are going in a direction where we will end up into specifying user-namespaces in two different ways.

May be you have noticed something in the current design which made you think so. We would really appreciate if you point that out and help us reaching a better design.

Reply:
> In future also, to enable pod (k8s namespace level) user-namespaces, this mapping will be read from runtime and then a sub-part will be used for each pod (or k8s namespace).

We might be getting ahead of ourselves with this here but for the per-Pod user namespace scenarios, why would the Container Runtime userns mapping defaults play any role? I view it as a default the setups falls back to when no uid/gid mapping will be provided with the Pod (via Pod Spec or from Kubernetes namespace default). But if explicit mapping is provided, it should be able to map to any host ids it wants, unrestricted by the Container Runtime default settings.